runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java

```diff
@@ -0,0 +1,625 @@
+      add(component);
+      handleLogic();
+    }
+  }
+
+  private void populate()
+  {
+    this.offerImages = new ImageIcon[8];
+    this.imageLabels = new JLabel[8];
+    this.itemNameLabels = new JLabel[8];
+    this.offerPriceLabels = new JLabel[8];
+    this.statusIndicators = new JLabel[8];
+    this.progressBars = new JProgressBar[8];
+
+    //TODO: via java 8 stream?
+    for (int i = 0; i < offerImages.length; i++) {
+      offerImages[i] = new ImageIcon(ItemManager.getImage(0)); //the icon (dwarf remains)
+      imageLabels[i] = new JLabel(); //holds icon
+      itemNameLabels[i] = new JLabel(); //holds item name
+      offerPriceLabels[i] = new JLabel(); //holds offer price
+      statusIndicators[i] = new JLabel(); //holds quantity bought/sold
+      progressBars[i] = new JProgressBar();
+    }
+
+  }
+
+  private void handleLogic()
+  {
+    ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+    ses.scheduleWithFixedDelay(() -> updateUI.accept(client.getGrandExchangeOffers()), 0, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  private Consumer<GrandExchangeOffer[]> updateUI = (offers) ->
+      Arrays.stream(offers)
+          .filter(Objects::nonNull)
+          .forEach(offer -> {
+
+            int idx = Arrays.asList(offers).indexOf(offer);
+
+            offerImages[idx] = new ImageIcon(
+                client.getGrandExchangeOffers() == null || client.getGrandExchangeOffers()[idx] == null
+                    ? ItemManager.getImage(0)
+                    : ItemManager.getImage(client.getGrandExchangeOffers()[idx].getItemId())
+            );
+
+            //TODO: "empty slot" image [instead of dwarf remains image]?
+            imageLabels[idx].setIcon(offer.getItemId() == 0 ? new ImageIcon(ItemManager.getImage(0)) : offerImages[idx]);
+
+            itemNameLabels[idx].setText(offer.getItemId() == 0 ? "Empty" : offer.getTotalQuantity() + "x " + client.getItemDefinition(offer.getItemId()).getName());
+
+            offerPriceLabels[idx].setText(ItemManager.quantityToStackSize(offer.getPrice()).endsWith("M")
+                || ItemManager.quantityToStackSize(offer.getPrice()).endsWith("K")
+                ? ItemManager.quantityToStackSize(offer.getPrice())
+                : ItemManager.quantityToStackSize(offer.getPrice()) + " gp");
+
+            statusIndicators[idx].setText(getIsBuySell(offer));
+
+            EventQueue.invokeLater(() -> progressBars[idx].setValue(offer.getTotalQuantity() == 0 ? 0 : (offer.getQuantitySold() / offer.getTotalQuantity()) * 100));
+
+            progressBars[idx].setStringPainted(true);
+            progressBars[idx].setBackground(decodeColour(progressBars[idx]));
+            progressBars[idx].setForeground(Color.white);
+            progressBars[idx].setString(offer.getQuantitySold() + " / " + offer.getTotalQuantity());
+
+          });
+
+  private String getIsBuySell(GrandExchangeOffer offer)
+  {
+    switch (offer.getProgress()) {
+      case 2:
```

Comment:
These constants would be good to place into GrandExchangeOffer in runelite-api, maybe with some javadoc.